packages/xdl/src/Project.ts

```diff
@@ -892,26 +902,87 @@ export async function publishAsync(
         'generated',
         'AppConstants.java'
       );
-      await ExponentTools.deleteLinesInFileAsync(
-        `START EMBEDDED RESPONSES`,
-        `END EMBEDDED RESPONSES`,
-        constantsPath
-      );
-      await ExponentTools.regexFileAsync(
-        '// ADD EMBEDDED RESPONSES HERE',
-        `
-        // ADD EMBEDDED RESPONSES HERE
-        // START EMBEDDED RESPONSES
-        embeddedResponses.add(new Constants.EmbeddedResponse("${fullManifestUrl}", "assets://shell-app-manifest.json", "application/json"));
-        embeddedResponses.add(new Constants.EmbeddedResponse("${androidManifest.bundleUrl}", "assets://shell-app.bundle", "application/javascript"));
-        // END EMBEDDED RESPONSES`,
-        constantsPath
-      );
-      await ExponentTools.regexFileAsync(
-        /RELEASE_CHANNEL = "[^"]*"/,
-        `RELEASE_CHANNEL = "${options.releaseChannel}"`,
-        constantsPath
-      );
+      if (fs.existsSync(constantsPath)) {
+        // This is an ExpoKit app
+        // We need to add EmbeddedResponse instances on Android to tell the runtime
+        // that the shell app manifest and bundle is packaged.
+        await ExponentTools.deleteLinesInFileAsync(
+          `START EMBEDDED RESPONSES`,
+          `END EMBEDDED RESPONSES`,
+          constantsPath
+        );
+        await ExponentTools.regexFileAsync(
+          '// ADD EMBEDDED RESPONSES HERE',
+          `
+          // ADD EMBEDDED RESPONSES HERE
+          // START EMBEDDED RESPONSES
+          embeddedResponses.add(new Constants.EmbeddedResponse("${fullManifestUrl}", "assets://shell-app-manifest.json", "application/json"));
+          embeddedResponses.add(new Constants.EmbeddedResponse("${androidManifest.bundleUrl}", "assets://shell-app.bundle", "application/javascript"));
+          // END EMBEDDED RESPONSES`,
+          constantsPath
+        );
+        await ExponentTools.regexFileAsync(
+          /RELEASE_CHANNEL = "[^"]*"/,
+          `RELEASE_CHANNEL = "${options.releaseChannel}"`,
+          constantsPath
+        );
+      } else {
+        // This is an app with expo-updates installed, so set the applicable meta-data properties in
+        // AndroidManifest.xml
+        let androidManifestXmlPath = path.join(
+          projectRoot,
+          'android',
+          'app',
+          'src',
+          'main',
+          'AndroidManifest.xml'
+        );
+        let androidManifestXmlFile = fs.readFileSync(androidManifestXmlPath, 'utf8');
+        let expoUpdateUrlRegex = /<meta-data android:name="expo.modules.updates.EXPO_UPDATE_URL" android:value="[^"]*" \/>/;
+        let expoSdkVersionRegex = /<meta-data android:name="expo.modules.updates.EXPO_SDK_VERSION" android:value="[^"]*" \/>/;
+        let expoReleaseChannelRegex = /<meta-data android:name="expo.modules.updates.EXPO_RELEASE_CHANNEL" android:value="[^"]*" \/>/;
```

Comment:
Modifying the Android manifest using regular expressions like these seems very prone to error. Because the manifest is also modified by other tools, even a line break in the middle of one of these tags or a different order of XML attributes could break this, even though both are valid in XML.
 
Could we adopt a different approach that makes use of an XML parser instead of regexes to make these modifications in the file? You could look in https://github.com/expo/expo-cli/blob/4472a15a54468be7dae3e477d2dce7c7be21f60e/packages/config/src/android/Manifest.ts for an existing example of this in this codebase.

Reply:
Thanks for pointing this out. I tried 2 different libraries for parsing and writing xml files -- xml2js (which is used in @expo/config) and [xml-js](https://github.com/nashwaan/xml-js), which is more fully featured and offers a non-compact mode. This work is visible in the [`@eric/expo-updates-wip`](https://github.com/expo/expo-cli/tree/%40eric/expo-updates-wip) branch.

Both approaches had issues -- in particular, parsing and rewriting the entire AndroidManifest.xml file created a much larger diff than I wanted. xml2js removed comments and changed the order of elements. xml-js was better, but despite offering some customization options, made many whitespace changes to the file. I suppose this is only ever an issue the first time someone runs this command, but I suspect it will affect almost everyone, since the default Android Studio XML style is not possible to achieve with this library. I could not find another library for building XML from JS with more whitespace customization options.

My opinion is that since this is supposed to be a convenient side-effect of running a command with an entirely different purpose (`expo publish`), it's better for us to have the brittle regexes that sometimes fail than to unintentionally create large and confusing diffs in developers' project files, which feels icky to me. Developers can always change these settings themselves (and there is clear documentation on how to do so). I did make a change to the regexes to make them more forgiving in terms of whitespace and attribute order. Let me know what you think!

Reply:
That seems like an acceptable tradeoff for now. If we at some point end up developing a tool to update AndroidManifest.xml without regexes, this code is quite clear and should be easy to adapt to use that tool. (I suspect that `Manifest.ts`, given that it’s also using xml2js, might also be suffering from some of the problems you mentioned, so we might visit this in the future.)